core/src/utils/sync/racy_lock.rs

```diff
@@ -0,0 +1,174 @@
+use alloc::boxed::Box;
+use core::{
+    fmt,
+    ops::Deref,
+    ptr,
+    sync::atomic::{AtomicPtr, Ordering},
+};
+
+/// Thread-safe, non-blocking, "first one wins" flavor of `once_cell::sync::OnceCell`
+/// with the same interface as `std::sync::LazyLock`.
+///
+/// The underlying implementation is based on `once_cell::sync::race::OnceBox` which relies on
+/// `core::atomic::AtomicPtr` to ensure that the data race results in a single successful
+/// write to the relevant pointer, namely the first write.
+/// See <https://github.com/matklad/once_cell/blob/v1.19.0/src/race.rs#L294>.
+///
+/// Performs lazy evaluation and can be used for statics.
+pub struct RacyLock<T, F = fn() -> T>
+where
+    F: Fn() -> T,
+{
+    inner: AtomicPtr<T>,
+    f: F,
+}
+
+impl<T, F> RacyLock<T, F>
+where
+    F: Fn() -> T,
+{
+    /// Creates a new lazy, racy value with the given initializing function.
+    pub const fn new(f: F) -> Self {
+        Self {
+            inner: AtomicPtr::new(ptr::null_mut()),
+            f,
+        }
+    }
+
+    /// Forces the evaluation of the locked value and returns a reference to
+    /// the result. This is equivalent to the `Deref` impl, but is explicit.
+    ///
+    /// This method will not block the calling thread if another initialization
+    /// routine is currently running.
+    pub fn force(this: &RacyLock<T, F>) -> &T {
+        let mut ptr = this.inner.load(Ordering::Acquire);
+
+        // Pointer is not yet set, attempt to set it ourselves.
+        if ptr.is_null() {
+            // Execute the initialization function and allocate.
+            let val = (this.f)();
+            ptr = Box::into_raw(Box::new(val));
+
+            // Attempt atomic store.
+            let exchange = this.inner.compare_exchange(
+                ptr::null_mut(),
+                ptr,
+                Ordering::AcqRel,
+                Ordering::Acquire,
+            );
+
+            // Pointer already set, load.
+            if let Err(old) = exchange {
+                drop(unsafe { Box::from_raw(ptr) });
```

Comment:
@plafer I did have this comment at the top:
```rust
/// The underlying implementation is based on `once_cell::sync::race::OnceBox` which relies on
/// `core::atomic::AtomicPtr` to ensure that the data race results in a single successful
/// write to the relevant pointer, namely the first write.
/// See <https://github.com/matklad/once_cell/blob/v1.19.0/src/race.rs#L294>.
```

Which indicates the whole implementation is derived from that source. Just checking - do we want to also repeat the same notion again for each function?

Reply:
Since `OnceBox` didn't have a `force()` method, I assumed you had figured out how to implement it yourself, only to later realize that our `force()` and `OnceBox`'s `get_try_or_init` are very similar. So I would add just a one-liner to point people in the right direction